viewing_party/party.py

```diff
@@ -1,23 +1,231 @@
 # ------------- WAVE 1 --------------------
 
 def create_movie(title, genre, rating):
-    pass
+
+    #return None of title, genre, and rating are empty
+    if not (title and genre and rating):
+        return None
+    
+    movie = {
+                "title": title,
+                "genre": genre,
+                "rating": rating
+            
+            }
+    return movie
+
+
+def add_to_watched(user_data, movie):
+
+    user_data["watched"].append(movie)
+    
+    return user_data
+
+
+def add_to_watchlist(user_data, movie):
+
+    user_data["watchlist"].append(movie)
+    
+    return user_data
+
+
+def watch_movie(user_data, title): 
+
+    #loop through each movie in the watchlist
+    for movie in user_data["watchlist"]:
+
+        #if it matches the title passed as the arg add it to watch
+        if title == movie["title"]:
+
+            user_data["watched"].append(movie)
+            user_data["watchlist"].remove(movie)
+        
+    return user_data
+            
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+
+    watched_movies = user_data["watched"]
+
+    if watched_movies: 
+        
+        subtotal = 0
+
+        for movie in watched_movies:
+
+            subtotal += movie["rating"] #add up ratings
+        
+        avg_rating = subtotal/float(len(watched_movies)) #calculate avg
+        
+    else: #avg is 0 if no watched movies
+        avg_rating = 0
+
+    return avg_rating
+
+
+def get_most_watched_genre(user_data):
+
+    #list of dictionaries with watched movies
+    watched_movies = user_data["watched"]
+
+    if not watched_movies:
+
+        return None
+
+    #initialize dictionary to track genres and frequencies
+    watched_genres = {}
+
+    #calculate frequency
+    for movie in watched_movies:
+
+        if movie["genre"] in watched_genres:
+
+            watched_genres[movie["genre"]] += 1
+        
+        else: 
+            watched_genres[movie["genre"]] = 1
+
+    
+    #find the frequency of most watched genre
+    
+    list_of_frequencies = list(watched_genres.values())
+
+    list_of_frequencies.sort()
+
+    most_watched_value = list_of_frequencies[-1]
+
+    #loop through the gengand find the first movie that matches that frequency
+    for genre in watched_genres:
+
+        if watched_genres[genre] == most_watched_value:
+
+            most_watched_genre = genre

```

Comment:
Once we have a frequency map, another approach would be to use the `max` function :
```suggestion
    most_watched_genre = max(watched_genres, key = watched_genres.get)
```